### PR TITLE
Perform gradient clipping on global batch when using gradient accumulation

### PR DESCRIPTION
**Pull request by ashors1**
Refactoring to allow gradient clipping to be performed on full batch rather than subbatches when using `ShardedStaticAccumulator`. Note that this refactor allows us to maintain support for `enable_skip_step_on_gradient_anomalies` and requires `x+1` grad norm calculations per global batch when using `ShardedStaticAccumulator` with `x` subbatches (once per subbatch to determine whether step should be skipped, once when applying gradient clipping in base optimizer update) and requires one grad clip per global batch.

This PR should be taken together with the corresponding [Praxis PR](https://github.com/google/praxis/pull/6).

### CODE REVIEW

**Comment by zhangqiaorjc** [comment]
@ashors1 sorry for the late review, could rebase to head? i want to import it and run some internal CI, thanks!

**Comment by zhangqiaorjc** [comment]
There's quite a few redundant whitespaces. Could you run some python linter to remove those?

File: paxml/learners.py [hunk 1]
```diff
@@ -335,7 +369,15 @@ def update_states(
     """
     p = self._hparams
 
-    grads, valid_step = self.scale_gradients(grads)
+    grad_norm, valid_step = self.get_grad_norm_valid_step(grads)
+
+    using_ga = hasattr(p.optimizer, 'num_sub_batches')
```
Comment (zhangqiaorjc) [inline paxml/learners.py:374]: nit: let's use `using_grad_accum`

most readers might not know what `ga` means

File: paxml/learners.py [hunk 2]
```diff
@@ -588,8 +631,16 @@ def scale_gradients_by_optimizer(
   ) -> Tuple[NestedMap, JTensor]:
     optimizer_mask, default_mask = self.get_masks(var_weight_hparams)
 
-    all_grads, all_valid_step = self.scale_gradients(
-        jax.tree_map(lambda x, y: x * y, raw_grads, default_mask),
+    raw_grads = jax.tree_map(lambda x, y: x * y, raw_grads, default_mask)
```
Comment (zhangqiaorjc) [inline paxml/learners.py:634]: let's not reuse `raw_grads`, let's call this `grads_after_mask` because you've introduced a subtle bug here if you look at line line 659 inside the auxiliary_optimizers loop, you are now combining this outer mask with inner mask

i would not overwrite `raw_grads` variable, just 

```
grads_after_mask = jax.tree_map(lambda x, y: x * y, raw_grads, default_mask)
grad_norm, all_valid_step = self.get_grad_norm_valid_step(
        grads_after_mask,
        optimizer_name='main',
    )
```

so that inside auxiliary_optimizers loop, `raw_grads` is only added to each auxiliary optimizer mask

**Review by zhangqiaorjc — CHANGES_REQUESTED** [review]

**Review by zhangqiaorjc — APPROVED** [review]

**Comment by nluehr** [comment]
@zhangqiaorjc is there a reason this has been approved by not merged yet?
